src/android/com/outsystems/firebase/analytics/OSFANLManager.kt

```diff
@@ -0,0 +1,162 @@
+import com.outsystems.firebase.analytics.validator.OSFANLEventItemsValidator
+import com.outsystems.firebase.analytics.validator.OSFANLEventParameterValidator
+import com.outsystems.firebase.analytics.validator.OSFANLMinimumRequired.AT_LEAST_ONE
+import com.outsystems.firebase.analytics.validator.OSFANLMinimumRequired.ONE
+import org.json.JSONArray
+import org.json.JSONObject
+
+class OSFANLManager {
+
+    @Throws(OSFANLError::class)
+    fun buildOutputEventFromInputJSON(input: JSONObject): OSFANLEventOutputModel {
+
+        val eventName = input.getString(EVENT.json)
+        val parameters = JSONArray(input.getString(EVENT_PARAMETERS.json))
+        val items = JSONArray(input.getString(ITEMS.json))
+
+        val eventBuilderMethod = getEventBuilderMethod(eventName)
+        val eventBundle = eventBuilderMethod(this, parameters, items)
+
+        return OSFANLEventOutputModel(eventName, eventBundle)
+    }
+
+    private fun getEventBuilderMethod(
+        eventName: String
+    ): (OSFANLManager, JSONArray, JSONArray) -> Bundle {
+
+        when (OSFANLAnalyticsEvent.fromValue(eventName)) {
+
+            null -> throw OSFANLError.unexpected(eventName)
+
+            OSFANLAnalyticsEvent.PURCHASE ->
+                return OSFANLManager::buildBundleForPurchaseParametersEventType
+
+            OSFANLAnalyticsEvent.REFUND ->
+                return OSFANLManager::buildBundleForRefundParametersEventType
+
+            OSFANLAnalyticsEvent.SELECT_ITEM ->
+                return OSFANLManager::buildBundleForOneItemParametersEventType
+
+            OSFANLAnalyticsEvent.SELECT_PROMOTION ->
+                return OSFANLManager::buildBundleForSelectPromotionParametersEventType
+
+            OSFANLAnalyticsEvent.VIEW_ITEM_LIST ->
+                return OSFANLManager::buildBundleForViewItemListParametersEventType
+
+            OSFANLAnalyticsEvent.VIEW_PROMOTION ->
+                return OSFANLManager::buildBundleForOneItemParametersEventType
+
+            else ->
+                return OSFANLManager::buildBundleForValueCurrencyAndMultipleItemsEventType
+        }
+    }
+
+    private fun buildBundleForValueCurrencyAndMultipleItemsEventType(
+        parameters: JSONArray,
+        items: JSONArray
+    ): Bundle {
+        return buildAndValidateParameterBundle(
+            parameters,
+            items,
+            OSFANLEventParameterValidator.Builder()
+                .requireCurrencyValue()
+                .build(),
+            OSFANLEventItemsValidator(AT_LEAST_ONE)
+        )
+    }
+
+    private fun buildBundleForPurchaseParametersEventType(
+        parameters: JSONArray,
+        items: JSONArray
+    ): Bundle {
+        return buildAndValidateParameterBundle(
+            parameters,
+            items,
+            OSFANLEventParameterValidator.Builder()
+                .requireCurrencyValue()
+                .required(TRANSACTION_ID)
+                .number(SHIPPING, TAX)
```

Comment:
Aren't we missing `VALUE` here?

Reply:
For the number requirement? That is being validated on the `.requireCurrencyValue`.

Reply:
It ensures the rule  "if one is present the other is required" and also validates value to be a number.

Reply:
Then why does the `buildBundleForRefundParametersEventType` method have a `.number(SHIPPING, TAX, VALUE)`, different from this one? Shouldn't the validation be the same in both methods?

Reply:
You're right. It's inconsistent.
I'll remove the number validation from there.